File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        // if(nums[0] == target) return 0;
        int l=0, r=n-1;
        while(l <= r){
            int mid = (l+r)/2;
            if(nums[mid] == target) return mid;
            if(nums[0] <= nums[mid]){
                if(target >= nums[0] && target <= nums[mid]) r = mid-1;
                else l = mid+1;
            } else {
                if(target >= nums[mid] && target <= nums[n-1]) l = mid+1;
                else r = mid-1;
            }
        }
        return -1;
    }
};

File: 33-search-in-rotated-sorted-array/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        n = len(nums)

        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            elif nums[0] <= nums[m]:
                if nums[0] <= target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if nums[m] < target <= nums[n - 1]:
                    l = m + 1
                else:
                    r = m - 1

        return -1